domain/machine/service/service.go

```diff
@@ -107,22 +123,52 @@ func NewService(st State) *Service {
 }
 
 // CreateMachine creates the specified machine.
+// It returns a MachineAlreadyExists error if a machine with the same name
+// already exists.
 func (s *Service) CreateMachine(ctx context.Context, machineName machine.Name) (string, error) {
 	// Make a new UUIDs for the net-node and the machine.
 	// We want to do this in the service layer so that if retries are invoked at
 	// the state layer we don't keep regenerating.
-	nodeUUID, err := uuid.NewUUID()
+	nodeUUID, machineUUID, err := createUUIDs()
 	if err != nil {
 		return "", errors.Annotatef(err, "creating machine %q", machineName)
 	}
-	machineUUID, err := uuid.NewUUID()
+
+	err = s.st.CreateMachine(ctx, machineName, nodeUUID, machineUUID)
+
+	return machineUUID, errors.Annotatef(err, "creating machine %q", machineName)
+}
+
+// CreateMachineWirhParent creates the specified machine with the specified
+// parent.
+// It returns a MachineAlreadyExists error if a machine with the same name
+// already exists.
+// It returns a NotFound error if the parent machine does not exist.
+func (s *Service) CreateMachineWithParent(ctx context.Context, machineName, parentName machine.Name) (string, error) {
```

Comment:
@gfouillet we should maybe do the check you were mentioning (regarding the grandparent machine) in this method right? I mean, the reboots is a feature of the machine(s) as we model them, so we shouldn't even have to check for a wrong modelling at that point, right?

Reply:
Yes it didn't come in my mind reading... But yes.

Check for grand parent should be done at two places:

* In my PR, in the reboot stuff, were they are required to the code be valid (and if for some reason this assumption change, we will be able to catch its impact quickly, since test would break)
* In this function `CreateMachineWithParent`, since it enforces the rule at core. Catching it in ShouldRebootOrShutdown is a last catch, but we shouldn't be in this situation anyway.

Reply:
`ShouldRebootOrShutdown` shouldn't check that, as satisfying no grandparent rule is outside its scope, it's a design concern that should be satisfied in spots we associate a parent with a machine. see https://github.com/juju/juju/pull/17759#discussion_r1683234662